src/slam_toolbox_common.cpp

```diff
@@ -63,49 +67,151 @@ void SlamToolbox::configure()
   for (int i = 0; i != threads_.size(); i++) {
+    threads_[i]->interrupt();
     threads_[i]->join();
+    threads_[i].reset();
   }
+  threads_.clear();
+  closure_assistant_.reset();
+  sst_->on_deactivate();
+  sstm_->on_deactivate();
+  pose_pub_->on_deactivate();
+  return CallbackReturn::SUCCESS;
+}
 
+/*****************************************************************************/
+CallbackReturn SlamToolbox::on_cleanup(const rclcpp_lifecycle::State &)
+/*****************************************************************************/
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
   smapper_.reset();
   dataset_.reset();
+  map_saver_.reset();
+  pose_helper_.reset();
+  laser_assistant_.reset();
+  scan_holder_.reset();
+  solver_.reset();
+
+  scan_filter_sub_.reset();
+  ssDesserialize_.reset();
+  ssSerialize_.reset();
+  ssPauseMeasurements_.reset();
+  ssMap_.reset();
+  sstm_.reset();
+  sst_.reset();
+  pose_pub_.reset();
+
+  tfB_.reset();
+  tfL_.reset();
+  tf_.reset();
+
+  lasers_.clear();
+  return CallbackReturn::SUCCESS;
+}
+
+/*****************************************************************************/
+CallbackReturn
+SlamToolbox::on_shutdown(const rclcpp_lifecycle::State & state)
+/*****************************************************************************/
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+  if (state.id() == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {
+    on_deactivate(state);
+    on_cleanup(state);
+  } else if (state.id() == lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE) {
+    on_cleanup(state);
+  }
+  RCLCPP_INFO(get_logger(), "Shutdown");
+  return CallbackReturn::SUCCESS;
+}
+
+/*****************************************************************************/
+SlamToolbox::~SlamToolbox()
+/*****************************************************************************/
+{
+  scan_filter_.reset();
   closure_assistant_.reset();
+
+  for (int i = 0; i != threads_.size(); i++) {
+    threads_[i]->interrupt();
+    threads_[i]->join();
+    threads_[i].reset();
+  }
+  threads_.clear();
+
+  smapper_.reset();
+  dataset_.reset();
   map_saver_.reset();
   pose_helper_.reset();
   laser_assistant_.reset();
   scan_holder_.reset();
   solver_.reset();
+
+  scan_filter_sub_.reset();
+  ssDesserialize_.reset();
+  ssSerialize_.reset();
+  ssPauseMeasurements_.reset();
+  ssMap_.reset();
+  sstm_.reset();
+  sst_.reset();
+  pose_pub_.reset();
+
+  tfB_.reset();
+  tfL_.reset();
+  tf_.reset();
+
+  lasers_.clear();
 }
 
 /*****************************************************************************/
 void SlamToolbox::setSolver()
 /*****************************************************************************/
 {
   // Set solver to be used in loop closure
-  std::string solver_plugin = std::string("solver_plugins::CeresSolver");
-  solver_plugin = this->declare_parameter("solver_plugin", solver_plugin);
-
+  std::string solver_plugin = this->get_parameter("solver_plugin").as_string();
```

Comment:
Is this guaranteed to already be declared?

Reply:
From below, doesn't look like it

Reply:
It is declared at the end of `setParams`. However, this is the effect when `setParams` was done in the constructor.  
Now `setParams` has been moved inside `on_configure`, so I'll update it so that it is declared just before.

https://github.com/SteveMacenski/slam_toolbox/blob/89d4a127459a934f444459a64eb6ec60b1102c4b/src/slam_toolbox_common.cpp#L333-L359